File: tools/calendar-bridge/calendar-bridge.swift
/**
 * calendar-bridge — EventKit subprocess for the OmniFocus MCP server.
 *
 * Per ADR-0018, this binary is the seam between Node and macOS EventKit.
 * The Node process spawns it with a subcommand on argv[1] and parses
 * one JSON line on stdout.
 *
 *   $ calendar-bridge ping
 *   {"ready":false,"reason":"awaiting-resource-implementation","permission":"granted"}
 *   $ calendar-bridge permission
 *   {"permission":"granted"}
 *   $ calendar-bridge request-access
 *   {"granted":true,"permission":"granted"}
 *
 * Output: one JSON line per invocation, written to stdout. Stderr receives
 * diagnostic messages (startup, errors). Stdout is strictly newline-delimited
 * JSON so the Node consumer never misparses.
 *
 * Subcommands:
 *
 *   ping           — health check + current authorization state. Read-only;
 *                    does NOT trigger the macOS Calendar TCC prompt.
 *   permission     — emit just the authorization state. Read-only; does NOT
 *                    trigger the prompt.
 *   request-access — request Calendar access. **Triggers the macOS TCC
 *                    prompt on first call** (when current state is
 *                    `not-determined`); subsequent calls return immediately
 *                    with the cached state. Async — the binary blocks on a
 *                    DispatchSemaphore until the EventKit completion handler
 *                    fires. Emits `{"granted": bool, "permission": "..."}`.
 *
 * `permission` values mirror EventKit's `EKAuthorizationStatus`:
 *   - "not-determined"  — never asked; `request-access` triggers the prompt
 *   - "denied"          — user denied; user must re-grant in System Settings
 *   - "restricted"      — denied at the OS level (parental controls, MDM)
 *   - "granted"         — permission granted; reads will succeed
 *
 * Subsequent slices of #484 add:
 *   - the `calendar` subcommand: read events in a [from, to] range
 *   - the `agenda` subcommand: merge calendar events with OF forecast data
 *
 * Lifecycle:
 *   - argv[1] = "ping"           → emit health + permission JSON, exit 0
 *   - argv[1] = "permission"     → emit just permission JSON, exit 0
 *   - argv[1] = "request-access" → trigger prompt (or return cached), exit 0
 *   - argv[1] = anything else    → diagnostic on stderr, exit 1
 *   - argv[1] missing            → diagnostic on stderr, exit 1
 *
 * Build:
 *   swiftc tools/calendar-bridge/calendar-bridge.swift -o bin/calendar-bridge-darwin-$(uname -m)
 *
 * @see docs/adr/0018-calendar-bridge-eventkit-only.md — architecture decision
 * @see scripts/build-calendar-bridge.sh — CI build script
 * @see tools/watcher/omnifocus-watcher.swift — sibling subprocess pattern
 */

import EventKit
import Foundation

let args = CommandLine.arguments
let argv1 = args.count > 1 ? args[1] : ""

func emit(_ json: String) {
    print(json)
}

func diagnostic(_ message: String) {
    FileHandle.standardError.write("calendar-bridge: \(message)\n".data(using: .utf8) ?? Data())
}

/// Map `EKAuthorizationStatus` to the wire-stable string the Node consumer
/// expects. `fullAccess` and `writeOnly` (macOS 14+) both map to `granted` for
/// our read-only purposes — we don't surface the distinction since this
/// bridge never writes (per ADR-0018 §3 read-only mutation policy).
func authorizationStatusString(_ status: EKAuthorizationStatus) -> String {
    switch status {
    case .notDetermined:
        return "not-determined"
    case .restricted:
        return "restricted"
    case .denied:
        return "denied"
    case .authorized, .fullAccess, .writeOnly:
        return "granted"
    @unknown default:
        return "not-determined"
    }
}

/// Trigger the macOS Calendar TCC prompt (or return cached state if already
/// answered) and emit the result. Uses `requestFullAccessToEvents` on
/// macOS 14+ and falls back to the deprecated `requestAccess(to:completion:)`
/// for older targets — both call the same TCC machinery underneath, so the
/// result is identical for our read-only use case.
///
/// The subprocess blocks on a `DispatchSemaphore` until the completion
/// handler fires, then exits. EventKit dispatches the callback on a
/// background thread, so we don't need a RunLoop pump — the semaphore is
/// sufficient to keep the main thread alive.
func requestAccessAndEmit() -> Never {
    let store = EKEventStore()
    let semaphore = DispatchSemaphore(value: 0)
    var grantedFlag = false

    let completion: (Bool, Error?) -> Void = { granted, error in
        grantedFlag = granted
        if let error = error {
            diagnostic("requestAccess error: \(error.localizedDescription)")
        }
        semaphore.signal()
    }

    if #available(macOS 14.0, *) {
        store.requestFullAccessToEvents(completion: completion)
    } else {
        store.requestAccess(to: .event, completion: completion)
    }

    semaphore.wait()

    // Re-query authorization status after the prompt resolves: `granted` is
    // the boolean result, but the EKAuthorizationStatus is the stable
    // surface we report alongside it.
    let resolved = authorizationStatusString(EKEventStore.authorizationStatus(for: .event))
    emit(#"{"granted":\#(grantedFlag),"permission":"\#(resolved)"}"#)
    exit(0)
}

let permission = authorizationStatusString(EKEventStore.authorizationStatus(for: .event))

switch argv1 {
case "ping":
    // Health check: report bridge readiness + current authorization state.
    // `ready` becomes `true` once the calendar/agenda subcommands ship; for
    // now the bridge is callable but the resources aren't wired up yet.
    let ready = false
    let reason = "awaiting-resource-implementation"
    emit(#"{"ready":\#(ready),"reason":"\#(reason)","permission":"\#(permission)"}"#)
    exit(0)

case "permission":
    emit(#"{"permission":"\#(permission)"}"#)
    exit(0)

case "request-access":
    requestAccessAndEmit()

case "":
    diagnostic("missing subcommand. Usage: calendar-bridge <subcommand>")
    diagnostic("subcommands: ping, permission, request-access")
    exit(1)

default:
    diagnostic("unknown subcommand: \(argv1)")
    diagnostic("subcommands: ping, permission, request-access")
    exit(1)
}

File: src/bridge/calendarBridge.ts
/**
 * `CalendarBridge` — Node-side TypeScript wrapper around the Swift
 * `calendar-bridge` binary (per ADR-0018).
 *
 * The Swift binary is the seam between Node and macOS EventKit. This wrapper
 * spawns it as a one-shot subprocess, parses one JSON line of stdout, and
 * surfaces typed errors so callers can handle the macOS TCC permission flow
 * and the missing-binary case without parsing strings.
 *
 *   const bridge = new CalendarBridge();
 *   const { permission } = await bridge.getPermission();
 *   const events = await bridge.readEvents(from, to);
 *
 * The wrapper is intentionally thin — it does not cache permission state
 * between calls (the user can revoke at any time) and does not retry on
 * failure. Callers that need higher-level orchestration (capabilities probe,
 * agenda merging) should layer above this.
 *
 * @see docs/adr/0018-calendar-bridge-eventkit-only.md — architecture decision
 * @see tools/calendar-bridge/calendar-bridge.swift — subprocess source
 * @see src/watcher/DatabaseWatcher.ts — sibling subprocess pattern
 */

import type { ChildProcess } from "node:child_process";
import { spawn as nodeSpawn } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { CalendarBridgeUnavailable, CalendarPermissionDenied } from "../errors/index.js";

// ---------------------------------------------------------------------------
// Wire-format types — match calendar-bridge.swift exactly
// ---------------------------------------------------------------------------

/** Authorization state surfaced by EventKit, mapped to wire-stable strings. */
export type CalendarPermission = "not-determined" | "denied" | "restricted" | "granted";

export interface PingResult {
  ready: boolean;
  reason: string;
  permission: CalendarPermission;
}

export interface PermissionResult {
  permission: CalendarPermission;
}

export interface RequestAccessResult {
  granted: boolean;
  permission: CalendarPermission;
}

export interface CalendarEvent {
  id: string;
  title: string;
  startsAt: string;
  endsAt: string;
  allDay: boolean;
  calendarName: string;
  calendarSource: string;
  location?: string;
  status: "confirmed" | "tentative" | "cancelled";
  isAttendee?: boolean;
}

// ---------------------------------------------------------------------------
// Spawn injection — tests stub this out so they don't depend on a built binary
// ---------------------------------------------------------------------------

/** Minimal subset of `child_process.spawn` we depend on, narrowed for tests. */
export type SpawnFn = (
  command: string,
  args: readonly string[],
  options: { env?: NodeJS.ProcessEnv },
) => ChildProcess;

export interface CalendarBridgeOptions {
  /**
   * Override the path to the `calendar-bridge` Swift binary.
   * Default: `<package-root>/bin/calendar-bridge`.
   */
  binaryPath?: string;
  /** Spawn implementation — overridden in tests. */
  spawn?: SpawnFn;
  /** `existsSync` implementation — overridden in tests. */
  existsSync?: (p: string) => boolean;
}

// ---------------------------------------------------------------------------
// Defaults
// ---------------------------------------------------------------------------

/**
 * Resolve the default binary path relative to this file's location so it
 * works identically in `src/` (tsx dev) and `dist/` (compiled) layouts:
 *   src/bridge/calendarBridge.ts  →  ../../bin/calendar-bridge
 *   dist/bridge/calendarBridge.js →  ../../bin/calendar-bridge
 */
function defaultBinaryPath(): string {
  const thisFile = fileURLToPath(import.meta.url);
  return path.resolve(path.dirname(thisFile), "../../bin/calendar-bridge");
}

// ---------------------------------------------------------------------------
// Implementation
// ---------------------------------------------------------------------------

export class CalendarBridge {
  private readonly binaryPath: string;
  private readonly spawn: SpawnFn;
  private readonly existsSync: (p: string) => boolean;

  constructor(options: CalendarBridgeOptions = {}) {
    this.binaryPath = options.binaryPath ?? defaultBinaryPath();
    this.spawn = options.spawn ?? (nodeSpawn as unknown as SpawnFn);
    this.existsSync = options.existsSync ?? fs.existsSync;
  }

  /** Health check; emits readiness flag and current authorization state. */
  async ping(): Promise<PingResult> {
    return this.invoke<PingResult>(["ping"]);
  }

  /** Current authorization state without triggering the macOS TCC prompt. */
  async getPermission(): Promise<PermissionResult> {
    return this.invoke<PermissionResult>(["permission"]);
  }

  /**
   * Trigger the macOS Calendar TCC prompt (or return cached state if already
   * answered). Blocks until EventKit's completion handler fires.
   */
  async requestAccess(): Promise<RequestAccessResult> {
    return this.invoke<RequestAccessResult>(["request-access"]);
  }

  /**
   * Read calendar events in the half-open interval `[from, to)`. Both arguments
   * are ISO-8601 strings with offset, e.g. `2026-04-29T00:00:00-05:00`.
   *
   * Throws `CalendarPermissionDenied` if the user has not granted access; the
   * caller can recover by invoking `requestAccess()` first.
   *
   * @param sources Optional comma-separated calendar-name substrings (matches
   *                the Swift binary's `OMNIFOCUS_CALENDAR_SOURCES` env var).
   */
  async readEvents(from: string, to: string, sources?: string): Promise<CalendarEvent[]> {
    const env = sources ? { ...process.env, OMNIFOCUS_CALENDAR_SOURCES: sources } : process.env;
    const result = await this.invoke<{ events: CalendarEvent[] } | CalendarErrorPayload>(
      ["calendar", from, to],
      env,
    );
    if ("error" in result) {
      if (result.error === "permission-denied") {
        throw new CalendarPermissionDenied({
          details: { permission: result.permission },
        });
      }
      throw new CalendarBridgeUnavailable(`calendar-bridge returned error: ${result.error}`, {
        details: { error: result.error },
      });
    }
    return result.events;
  }

  // ---------------------------------------------------------------------------
  // Internals
  // ---------------------------------------------------------------------------

  private async invoke<T>(args: readonly string[], env?: NodeJS.ProcessEnv): Promise<T> {
    if (!this.existsSync(this.binaryPath)) {
      throw new CalendarBridgeUnavailable(
        `calendar-bridge binary not found at ${this.binaryPath}`,
        { details: { binaryPath: this.binaryPath } },
      );
    }

    let proc: ChildProcess;
    try {
      proc = this.spawn(this.binaryPath, args, env ? { env } : {});
    } catch (cause) {
      throw new CalendarBridgeUnavailable(
        `failed to spawn calendar-bridge: ${(cause as Error).message}`,
        { cause, details: { binaryPath: this.binaryPath } },
      );
    }

    return new Promise<T>((resolve, reject) => {
      let stdout = "";
      let stderr = "";

      proc.stdout?.on("data", (chunk: Buffer | string) => {
        stdout += typeof chunk === "string" ? chunk : chunk.toString("utf8");
      });
      proc.stderr?.on("data", (chunk: Buffer | string) => {
        stderr += typeof chunk === "string" ? chunk : chunk.toString("utf8");
      });

      proc.on("error", (err) => {
        reject(
          new CalendarBridgeUnavailable(`calendar-bridge spawn error: ${err.message}`, {
            cause: err,
          }),
        );
      });

      proc.on("close", (code) => {
        if (code !== 0) {
          reject(
            new CalendarBridgeUnavailable(
              `calendar-bridge exited with code ${code}: ${stderr.trim()}`,
              { details: { exitCode: code, stderr: stderr.trim() } },
            ),
          );
          return;
        }
        const line = stdout.trim().split("\n")[0] ?? "";
        if (!line) {
          reject(
            new CalendarBridgeUnavailable("calendar-bridge produced no output", {
              details: { stderr: stderr.trim() },
            }),
          );
          return;
        }
        try {
          resolve(JSON.parse(line) as T);
        } catch (cause) {
          reject(
            new CalendarBridgeUnavailable(
              `calendar-bridge produced unparsable JSON: ${line.slice(0, 200)}`,
              { cause, details: { line } },
            ),
          );
        }
      });
    });
  }
}

interface CalendarErrorPayload {
  error: string;
  permission: CalendarPermission;
}
